### ISSUE TITLE
Instrumentation yields unexpected object to "rack.attack" subscriptions.

### ISSUE THREAD

**Issue by Domenoth** [kickstarter/rack-attack#255]
Based on the [ActiveSupport documentation](http://api.rubyonrails.org/classes/ActiveSupport/Notifications.html#module-ActiveSupport::Notifications-label-Subscribers):

```
ActiveSupport::Notifications.subscribe('render') do |name, start, finish, id, payload|
  name    # => String, name of the event (such as 'render' from above)
  start   # => Time, when the instrumented block started execution
  finish  # => Time, when the instrumented block ended execution
  id      # => String, unique ID for this notification
  payload # => Hash, the payload
end
```

The block parameter `payload` is required to be a `Hash`.

However, Rack::Attack [defines `instrument` as](https://github.com/Domenoth/rack-attack/blob/master/lib/rack/attack.rb#L99):

```
    def instrument(req)
      notifier.instrument('rack.attack', req) if notifier
    end
```

And it passes a `Rack::Attack::Request` object instead of a `Hash`.

This causes issues when using [`attach_to`](http://api.rubyonrails.org/classes/ActiveSupport/Subscriber.html#method-c-attach_to):

```
  1) Error:
Rack::Attack.instrument::with throttling#test_0001_should instrument without error:
NoMethodError: undefined method `merge!' for #<Rack::Attack::Request:0x00007fdc329b1598>
```

The `Hash` method `merge!` is called on `payload` as part of the [`ActiveSupport::Subscriber` internal method `finish`](https://github.com/rails/rails/blob/a43e7706be907e62fad4ffe5f80ab9532c51dbca/activesupport/lib/active_support/subscriber.rb#L96).

**Comment by grzuy** [comment]
Hi @Domenoth, thank you for reporting this. You're right, current implementation it's not following `ActiveSupport`s spec.

**Comment by grzuy** [comment]
Targeting 6.0 because this would be a breaking change for existing implementations.

**Comment by grzuy** [comment]
For what is worth, this issue was already discussed in https://github.com/kickstarter/rack-attack/issues/100.

**Comment by grzuy** [comment]
This is now fixed in `v6.0.0`.
